oak-run/src/main/js/oak-mongo.js

```diff
@@ -452,6 +452,70 @@ var oak = (function(global){
         }
     };
 
+    /**
+     * Removes all unmerged branches on the document with the given path and
+     * clusterId. This method will only remove unmerged branches when the
+     * clusterId is inactive.
+     *
+     * @memberof oak
+     * @method removeUnmergedBranches
+     * @param {string} path the path of a document
+     * @param {number} clusterId collision markers for this clusterId will be removed.
+     * @param {number} [limit=1000000] maximum number of unmerged branches to remove.
+     * @returns {object} the result of the MongoDB update.
+     */
+    api.removeUnmergedBranches = function(path, clusterId, limit) {
+        if (path === undefined) {
+            print("No path specified");
+            return;
+        }
+        if (clusterId === undefined) {
+            print("No clusterId specified");
+            return;
+        }
+        if (limit === undefined) {
+            limit = 1000000;
+        }
+        // refuse to remove when clusterId is marked active
+        var clusterNode = db.clusterNodes.findOne({_id: clusterId.toString()});
+        if (clusterNode && clusterNode.state == "ACTIVE") {
+            print("Cluster node with id " + clusterId + " is active!");
+            print("Can only remove unmerged branches for inactive cluster node.");
+            return;
+        }
```

Comment:
What I was always wondering is whether these utility methods that require inactive clusterNodes (this and the removeCollisions one) should actually acquire the lock and only proceed if that was successful.
* they could set themselves to `invisible` to avoid discovery interference
* they could set the lease timeout high enough, so that under most circumstances it is able to finish in that time without complicated lease update logic in this method (eg 5min?)
* they should release the lock upon finishing (successful or when failing)
* in case they'd fail it would leave the lock for as many minutes we've chosen the lease timeout to - but as it is only an invisible lock, that wouldn't really do much harm (but maybe we'd need a method to force-unlock too?)

The risk we currently have is that there is interference with an automatically triggered startup of a node (perhaps unknowingly/unnoticed) which happens to pick this clusterNodeId .. and depending on logic it could or could not be a problem..

If we'd have the lease-lock in these javascript methods, it might also allow to implement larger cleanups of many collisions/branch-commits (for which we currently go via `limit`)

Reply:
It's a good point. But even acquiring the clusterId is not 100% correct. A lease timeout of 5 minutes may not be enough depending on the time the remove operation takes. I'd rather keep this simple implement something production grade into oak-store-document or oak-run in Java. Otherwise more logic needs to be duplicated here in oak-mongo.js.

Reply:
+1, sounds good